File: py/horus/pb/project_manager/service_handler.py
import abc as _abc
from ...rpc.base_handler import BaseRpcHandler, RpcMessage, WebSocket

from ..project_manager.service_pb2 import GetHealthStatusRequest, GetHealthStatusResponse

class ProjectManagerServiceHandler(BaseRpcHandler, metaclass=_abc.ABCMeta):
    """
    Service which stores configuration for all other services and coordinates updates to this configuration.
    """

    service_full_name = "horus.pb.ProjectManagerService"
    service_id = 3

    def __init__(self, ws: WebSocket):
        super().__init__(ws, {
            38: (GetHealthStatusRequest, GetHealthStatusResponse, self.get_health_status),
        })

    @_abc.abstractmethod
    async def get_health_status(self, request: GetHealthStatusRequest) -> GetHealthStatusResponse:
        """
        Provides horus health indicators.
        """
        raise NotImplementedError

File: cpp/horus/pb/project_manager/service_handler.cpp
#include "horus/pb/project_manager/service_handler.h"

#include <utility>

namespace horus {
namespace sdk {
namespace pb {

AnyFuture<pb::RpcMessage> ProjectManagerServiceHandler::Handle(const RpcContext& context,
                                                               pb::RpcMessage&& request) noexcept(false) {
  const pb::RpcMessage owned_request{std::move(request)};
  switch (owned_request.method_id()) {
    case 38: {
      return DoDispatch(&ProjectManagerServiceHandler::GetHealthStatus, context, owned_request);
    }
    default: {
      throw horus_internal::UnknownRpcMethodError{};
    }
  }
}

}  // namespace pb
}  // namespace sdk
}  // namespace horus
